Implement layer caching in docker persona

jzt opened the pull request:
This implements a layer cache on the docker persona. Any time the image store is mutated, a call to `Update()` is made, which will add new layers to the cache (and eventually remove them once we implement `docker rmi`. When `docker images` is called, this cache is accessed within the persona to build a list of images to present to the docker client.

I started to add all the typical methods like `Get()`, `Delete()`, etc, but realized none of those can sensibly be called in the persona because there is no knowledge about the layer(s) available to the persona without specifically calling the portlayer's `ListImages()` function to get a full list.

Looking for feedback on the initial approach, so tagging @corrieb, @hickeng, @cgtexmex, @sflxn, @caglar10ur.

Fixes #778

jzt commented on lib/apiservers/engine/backends/layer_cache.go:
```diff
@@ -0,0 +1,98 @@
+	)
+
+	if err != nil {
+		if _, ok := err.(*storage.CreateImageStoreConflict); ok {
```
I probably should change this to just overwrite existing layers in the cache to cover the case of an image having been updated.

jzt commented on lib/apiservers/engine/backends/layer_cache.go:
```diff
@@ -0,0 +1,98 @@
+
+	for _, layer := range layers.Payload {
+		// TODO(jzt): This assumes that layer ID and their data will remain consistent going forward
+		if _, ok := c.cache[layer.ID]; !ok {
```
I probably should change this to just overwrite existing layers in the cache to cover the case of an image having been updated.

caglar10ur commented:
LGTM :sheep: :it:

we should document that this is not getting persistent so restart or crash will make system in a weird state

cgtexmex commented on lib/apiservers/engine/backends/image.go:
```diff
@@ -184,6 +155,7 @@ func (i *Image) PullImage(ctx context.Context, ref reference.Named, metaHeaders
 		return err
 	}
 
+	layerCache.Update()
```
Why do you update the layer cache on images which is a read function?  Why wouldn't we update images on the pull only?

jzt replied:
That's actually at the bottom of the `PullImages` function (I think your diff collapsed it making it appear like it was in `Images`).

cgtexmex replied:
Ahh..thanks!

sflxn replied:
Ok, since it doesn't show up on my view either, I will add the comment here.  You're doing a full update of the cache on every pull.  That seems rather excessive.  Can't we just get the metadata for that one image and save it to the cache?

jzt replied:
No, that's the problem with imagec, remember? The docker persona knows absolutely nothing about an image it just pulled other than the reference that was used to pull it, and nowhere in the storage API is there support for getting a single layer based on this information.

sflxn replied:
Yeah, I see the problem.  I guess once we have the persistent key/value store, we can optimize this.  Ok, the full update is a must for now.

cgtexmex commented on lib/apiservers/engine/backends/image.go:
```diff
@@ -67,44 +64,18 @@ func (i *Image) ImageHistory(imageName string) ([]*types.ImageHistory, error) {
-	for _, layer := range layers.Payload {
+	for _, layer := range layerCache.cache {
 		imageConfig := &metadata.ImageConfig{}
 		if err := json.Unmarshal([]byte(layer.Metadata["metaData"]), imageConfig); err != nil {
```
Why not have an image cache?  Then the marshaling could be done during cache update and not on read...

jzt replied:
This is actually a really good reason to have an Image cache. Between this and @sflxn's reasons on slack, I'm convinced. Update incoming.

sflxn replied:
Yes, I agree with Clint's comments.  Also, my earlier comments below elaborates why we shouldn't cache the blobs around since they're swagger blobs.

cgtexmex commented:
A general question -- why not create a package in the backends for the caches / stores to reside?  We could then refactor the container store into that cache / store location?

sflxn commented on lib/apiservers/engine/backends/vicbackends.go:
```diff
@@ -28,6 +29,9 @@ const (
 	portLayerServerAddr string
+
+	// TODO(jzt): should this be defined in backends/image.go?
+	layerCache = NewLayerCache()
```
I would really like to have some kind of a getter function to get this.  Since this file is separate but in the same package, addressing this via direct name might be confusing to some people because they don't know where this var is.  Same as we do with PortLayerClient.

fdawg4l replied:
I think this instantiation and usage is fine.

sflxn replied:
I'm referring not to the instantiation but to the fact that we're accessing layerCache directly, with the assumption that layerCache is in the same namespace.  When I first looked at the image.go code, I wondered where layerCache was declared.

sflxn commented:
I think we need to review the CLI and API and see if the personality server ever needs to access the layer info.  I cannot think of an instance, but if there is, then the cache is ok.  If we never need access to the layer, then we should not store layers in the cache.  Instead we should just store image data.  I'll repeat, we should research this need before we remove it.

sflxn commented:
We should add getter and setter to the cache.  If we lock the cache for update and then allow direct access to the underlying data structure, then the lock is not really doing anything.  Also, we shouldn't assume this cache will always remain in the vicbackends namespace.  If we ever decide to move it, we'll have to add these getters and setters anyway.

jzt commented:
@sflxn It's so trivial to add back that I think it may be a waste of time to comb through APIs looking for an outlying case that might need it. I'd rather just ditch it for now, and add it back if/when we find we're going to need it.

sflxn commented:
I would rather we do not cache a dictionary of Swagger based structs.  Why don't we unmarshall the data once into imageconfig and store it in the cache instead?  I see Swagger as part of the transport layer and the personality server needs to use it but shouldn't assume Swagger is there.  I know we have to have Swagger info around, but we need to strive and to keep the Swagger out of the backend.  Although, that is a longer term desire.

jzt commented:
@sflxn can you comment in the actual code so that I can reply to comments individually rather than trying to address them in this general conversation?

jzt commented:
@sflxn Agreed with not caching swagger structs. I think that falls in line with @cgtexmex's comment about unmarshalling in the image cache.

sflxn commented on lib/apiservers/engine/backends/layer_cache.go:
```diff
@@ -0,0 +1,94 @@
+	sync.Mutex
+
+	// cache maps layer ID to layer data
+	cache map[string]*models.Image
```
We really need getters and setters to this cache.  It's the only way we can lock the cache.  My earlier out non inline comments explains this more and also adds other reasons.

jzt replied:
I think I addressed this in the changes I just pushed. Can you take a gander and let me know if this is more in line with what you were thinking?

sflxn commented:
LGTM.  BTW, I do agree with @cgtexmex that it should be in another package, but we can reorg it later.  Also, apparently Go adherents likes their getters to be named a certain way, ImageCache() instead of GetImageCache(), but I could care less.

caglar10ur commented on lib/apiservers/engine/backends/image_cache.go:
```diff
@@ -0,0 +1,125 @@
+// by a call to the portlayer. It is updated when new layers are pulled or
+// layers are deleted.
+type ImageCache struct {
+	sync.Mutex
```
Don't embed mutex into a struct. This makes ImageCache to have Lock/Unlock methods on it, instead use 

```
type ImageCache struct {
    m sync.Mutex
...
}
```
to keep it internal to the struct

caglar10ur replied:
Alternatively, if you think read is the hot path, you can convert this to RWMutex and grab the write lock in Update and read lock in Get

jzt replied:
Fixed, and thanks for the heads up on `RWMutex`, I learned something new today. :)

sflxn commented on lib/apiservers/engine/backends/image_cache.go:
```diff
@@ -15,43 +15,48 @@
-	// cache maps layer ID to layer data
-	cache map[string]*models.Image
+	// cache maps image ID to image metadata
+	cache map[string]*types.Image
```
One comment.  We chatted about it over Slack, but I'll add it here for posterity.  types.Image doesn't have all the data we need.  ImageConfig does, but not types.Image.

caglar10ur commented:
LGTM :sheep: :it:

jzt commented:
@cgtexmex, @sflxn: I've addressed all your comments with my last change:
- now caching `ImageConfig` structs instead of `types.Image`
- refactored the cache into its own package
- changed `GetImageCache()` to `ImageCache()` (couldn't do that before because image cache lived in the `vicbackends` package
- Added a getter for images

jzt commented:
Rebased, will merge after successful build.

jzt commented:
♻️
